feat: Node snapshot sync

spalladino opened the pull request:
- Utility functions for managing snapshot indices and uploading and downloading from gcs.
- Exposed `mdb_copy` from native world state and lmdb store
- Added `backupTo` methods to archiver and world state
- Added `actions` to the aztec node for uploading and downloading snapshots

Fixes #12926

spalladino committed:
feat: Stdlib for chain snapshots

Utility functions for managing snapshot indices and uploading and
downloading from gcs.

spalladino committed:
Lint

spalladino committed:
Implement mdb_copy in native world state and lmdb store

spalladino committed:
Expose db copy in world state and archiver, and remove unused fork method

Maddiaa0 commented on yarn-project/prover-client/src/proving_broker/proof_store/gcs_proof_store.ts:
```diff
@@ -16,6 +16,7 @@ import type { ProofStore } from './proof_store.js';
 
 const INPUTS_PATH = 'inputs';
 
+// REFACTOR: Use the stdlib/file-store instead of referencing google-cloud-storage directly.
```
issue?

spalladino replied:
Will do!

Maddiaa0 commented on yarn-project/stdlib/src/snapshots/upload.ts:
```diff
@@ -0,0 +1,43 @@
+
+  const basePath = getBasePath(metadata);
+  const [archiverDataUrl, worldStateDataUrl] = await Promise.all([
+    store.upload(`${basePath}/archiver-${date}-${metadata.l2BlockHash}.db`, archiverPath, { compress: true }),
```
Is there a strong deason to have the date in the name, at a first glance it would make downloading more difficult

Maddiaa0 replied:
Ah i see there is one index.json that points to the most recent snapshot

Maddiaa0 reviewed (approved):
im no voice of authority on the cpp pinging @alexghr but the ts looks good

spalladino committed:
Wire snapshot upload

spalladino commented:
@Maddiaa0 I added commit "Wire snapshot upload" that completes the loop, feel free to review just that one.

spalladino commented:
This is what an `index.json` file looks like:

```bash
$ cat /tmp/aztec/aztec-31337-1-0x26db1dbde404abe656c354e6331217ea7c9cc21e/index.json 
```

```json
{
  "l1ChainId": 31337,
  "l2Version": 1,
  "rollupAddress": "0x26db1dbde404abe656c354e6331217ea7c9cc21e",
  "snapshots": [
    {
      "l1BlockNumber": 131,
      "l2BlockHash": "0x1ddc5c6b6464ac19f27e1d6ed86dab06d7b834a20d5a785825522c8d82d521c3",
      "l2BlockNumber": 31,
      "schemaVersions": {
        "archiver": 1,
        "worldState": 1
      },
      "timestamp": 1742940387189,
      "dataUrls": {
        "l1-to-l2-message-tree": "file:///tmp/aztec/aztec-31337-1-0x26db1dbde404abe656c354e6331217ea7c9cc21e/l1-to-l2-message-tree-20250325220627-0x1ddc5c6b6464ac19f27e1d6ed86dab06d7b834a20d5a785825522c8d82d521c3.db",
        "archive-tree": "file:///tmp/aztec/aztec-31337-1-0x26db1dbde404abe656c354e6331217ea7c9cc21e/archive-tree-20250325220627-0x1ddc5c6b6464ac19f27e1d6ed86dab06d7b834a20d5a785825522c8d82d521c3.db",
        "public-data-tree": "file:///tmp/aztec/aztec-31337-1-0x26db1dbde404abe656c354e6331217ea7c9cc21e/public-data-tree-20250325220627-0x1ddc5c6b6464ac19f27e1d6ed86dab06d7b834a20d5a785825522c8d82d521c3.db",
        "note-hash-tree": "file:///tmp/aztec/aztec-31337-1-0x26db1dbde404abe656c354e6331217ea7c9cc21e/note-hash-tree-20250325220627-0x1ddc5c6b6464ac19f27e1d6ed86dab06d7b834a20d5a785825522c8d82d521c3.db",
        "nullifier-tree": "file:///tmp/aztec/aztec-31337-1-0x26db1dbde404abe656c354e6331217ea7c9cc21e/nullifier-tree-20250325220627-0x1ddc5c6b6464ac19f27e1d6ed86dab06d7b834a20d5a785825522c8d82d521c3.db",
        "archiver": "file:///tmp/aztec/aztec-31337-1-0x26db1dbde404abe656c354e6331217ea7c9cc21e/archiver-20250325220627-0x1ddc5c6b6464ac19f27e1d6ed86dab06d7b834a20d5a785825522c8d82d521c3.db"
      }
    },
    {
      "l2BlockNumber": 24,
      "l2BlockHash": "0x06e938f0668bb121838607e573d3aa21b0a0ed3a247655de5849887ea7bc5425",
      "l1BlockNumber": 117,
      "timestamp": 1742940331978,
      "schemaVersions": {
        "archiver": 1,
        "worldState": 1
      },
      "dataUrls": {
        "l1-to-l2-message-tree": "file:///tmp/aztec/aztec-31337-1-0x26db1dbde404abe656c354e6331217ea7c9cc21e/l1-to-l2-message-tree-20250325220531-0x06e938f0668bb121838607e573d3aa21b0a0ed3a247655de5849887ea7bc5425.db",
        "archive-tree": "file:///tmp/aztec/aztec-31337-1-0x26db1dbde404abe656c354e6331217ea7c9cc21e/archive-tree-20250325220531-0x06e938f0668bb121838607e573d3aa21b0a0ed3a247655de5849887ea7bc5425.db",
        "public-data-tree": "file:///tmp/aztec/aztec-31337-1-0x26db1dbde404abe656c354e6331217ea7c9cc21e/public-data-tree-20250325220531-0x06e938f0668bb121838607e573d3aa21b0a0ed3a247655de5849887ea7bc5425.db",
        "note-hash-tree": "file:///tmp/aztec/aztec-31337-1-0x26db1dbde404abe656c354e6331217ea7c9cc21e/note-hash-tree-20250325220531-0x06e938f0668bb121838607e573d3aa21b0a0ed3a247655de5849887ea7bc5425.db",
        "nullifier-tree": "file:///tmp/aztec/aztec-31337-1-0x26db1dbde404abe656c354e6331217ea7c9cc21e/nullifier-tree-20250325220531-0x06e938f0668bb121838607e573d3aa21b0a0ed3a247655de5849887ea7bc5425.db",
        "archiver": "file:///tmp/aztec/aztec-31337-1-0x26db1dbde404abe656c354e6331217ea7c9cc21e/archiver-20250325220531-0x06e938f0668bb121838607e573d3aa21b0a0ed3a247655de5849887ea7bc5425.db"
      }
    }
  ]
}
```

spalladino committed:
Add fast-sync action to server (pending wiring)

alexghr commented on yarn-project/aztec-node/src/actions/fast-sync.ts:
```diff
@@ -0,0 +1,161 @@
+// Half day worth of L1 blocks
+const MIN_L1_BLOCKS_TO_TRIGGER_REPLACE = 86400 / 2 / 12;
+
+export async function tryFastSync(
```
This seems to be uncalled?

spalladino replied:
Should've said it was WIP. It's now called on the node's `createAndSync`, right before initializing the archiver..

alexghr reviewed (commented):
Code looks good to me. Just a couple of questions around tightening up access to the snapshot mechanism

alexghr commented on yarn-project/stdlib/src/snapshots/upload.ts:
```diff
@@ -0,0 +1,49 @@
+  };
+  snapshotsIndex.snapshots.unshift(newSnapshotMetadata);
+
+  await store.save(getSnapshotIndexPath(metadata), Buffer.from(jsonStringify(snapshotsIndex, true)));
```
A bit worried if this fails (pod dies or network issues) then we'd left with an invalid index file.

spalladino replied:
I understand gcs ensures we don't end up with broken objects if an upload is aborted halfway. And the local store is just used for testing.

alexghr commented on yarn-project/archiver/src/archiver/kv_archiver_store/kv_archiver_store.ts:
```diff
@@ -49,6 +53,15 @@ export class KVArchiverDataStore implements ArchiverDataStore {
     this.#contractInstanceStore = new ContractInstanceStore(db);
   }
 
+  public async backupTo(path: string, compress = true): Promise<string> {
+    await this.db.backupTo(path, compress);
+    return join(path, 'data.mdb');
+  }
+
```
Forgot to go back and update this comment: this is handled already. This info is part of the snapshot index

spalladino replied:
I'm adding it to the snapshot metadata. We could write those as entries of the db itself though, and completely do away with the `db_version` file we use today. WDYT?

spalladino committed:
Fix tsc build issue

spalladino committed:
Fix bad merge

spalladino committed:
Wire fast sync and add e2e test

spalladino commented on yarn-project/aztec-node/src/aztec-node/server.ts:
```diff
@@ -991,6 +992,38 @@ export class AztecNodeService implements AztecNode, AztecNodeAdmin, Traceable {
     return this.validatorsSentinel?.computeStats() ?? Promise.resolve({ stats: {}, slotWindow: 0 });
   }
 
+  public async startSnapshotUpload(location: string): Promise<void> {
```
Yep, it's admin API, and requires permissions on the remote file store. So we should be the only ones allowed to call it for our nodes and our bucket.

spalladino commented on yarn-project/archiver/src/archiver/archiver.ts:
```diff
@@ -542,6 +542,18 @@ export class Archiver extends EventEmitter implements ArchiveSource, Traceable {
   }
 
+  /** Restarts the archiver after a stop. */
+  public restart() {
```
Good point, renamed to `resumeSync`

spalladino commented on yarn-project/aztec-node/src/actions/fast-sync.ts:
```diff
@@ -0,0 +1,161 @@
+    }
+  } finally {
+    log.info(`Cleaning up download dir ${downloadDir}`);
+    await rm(downloadDir, { recursive: true, maxRetries: 3 });
```
Added a `tryRmDir` helper

spalladino commented on yarn-project/aztec-node/src/actions/fast-sync.ts:
```diff
@@ -0,0 +1,161 @@
+  const downloadDir = await mkdtemp(join(dataDirectory, 'download-'));
+  try {
+    // Green light. Download the snapshot to a temp location.
+    const downloadPaths = makeSnapshotLocalPaths(downloadDir);
+    log.info(`Downloading snapshot from ${snapshotsUrl} to ${downloadDir}`, { snapshot, downloadPaths });
+    await downloadSnapshot(snapshot, downloadPaths, fileStore);
+    log.info(`Snapshot downloaded at ${downloadDir}`, { snapshotsUrl, snapshot, downloadPaths });
```
Good call, I'll add it

spalladino replied:
Created https://github.com/AztecProtocol/aztec-packages/issues/13068 for now

spalladino commented on yarn-project/aztec-node/src/aztec-node/server.ts:
```diff
@@ -991,6 +992,38 @@ export class AztecNodeService implements AztecNode, AztecNodeAdmin, Traceable {
+    }
+
+    // Test that the archiver has done an initial sync.
+    try {
+      archiver.getL1BlockNumber();
+    } catch (err) {
+      throw new Error(`Archiver initial sync not complete. Cannot start snapshot.`);
+    }
```
Good point, fixed!

spalladino commented on yarn-project/aztec-node/src/aztec-node/server.ts:
```diff
@@ -991,6 +992,38 @@ export class AztecNodeService implements AztecNode, AztecNodeAdmin, Traceable {
+    }
+
+    // Do not wait for the upload to be complete to return to the caller
+    void uploadSnapshot(
+      location,
+      this.blockSource as Archiver,
+      this.worldStateSynchronizer,
+      this.config,
+      this.log,
+    ).catch(err => this.log.error(`Error uploading snapshot: ${err}`));
+    return Promise.resolve();
```
A `isUploadingSnapshot` boolean variable is easier :-P

spalladino commented on yarn-project/aztec/src/cli/aztec_start_options.ts:
```diff
@@ -273,6 +273,19 @@ export const aztecStartOptions: { [key: string]: AztecStartOption[] } = {
+      flag: '--node.syncMode <value>',
+      description:
+        'Set sync mode to `full` to always sync via L1, `fast` to download a snapshot if there is no local data, `replace` to download even if there is local data.',
+      defaultValue: 'fast',
```
I think `fast` is good enough, since it only kicks in if there is no local data, so it does not overwrite anything. But I can be persuaded otherwise.

spalladino commented on yarn-project/stdlib/src/file-store/factory.ts:
```diff
@@ -0,0 +1,38 @@
+  } else if (config.startsWith('gs://')) {
+    try {
+      const url = new URL(config);
+      const bucket = url.host;
```
This is loaded from config, so we should be good

spalladino commented on yarn-project/stdlib/src/file-store/factory.ts:
```diff
@@ -0,0 +1,38 @@
+      const bucket = url.host;
+      const path = url.pathname.replace(/^\/+/, '');
+      logger.info(`Creating google cloud file store at ${bucket} ${path}`);
+      return new GoogleCloudFileStore(bucket, path);
```
Added a checkCredentials method

spalladino commented on yarn-project/stdlib/src/file-store/gcs.ts:
```diff
@@ -0,0 +1,87 @@
+    try {
+      const bucket = this.storage.bucket(this.bucketName);
+      const file = bucket.file(fullPath);
+      await file.save(data);
```
Yep

spalladino commented on yarn-project/stdlib/src/file-store/gcs.ts:
```diff
@@ -0,0 +1,87 @@
+        destination: file,
+        gzip: opts.compress,
+      };
+      await bucket.upload(srcPath, uploadOpts);
```
I believe this is not controlled by non-admin users

spalladino committed:
Apply suggestions from review and rename

spalladino commented on yarn-project/stdlib/src/file-store/local.ts:
```diff
@@ -0,0 +1,46 @@
+      .catch(() => false);
+  }
+
+  private getFullPath(pathOrUrl: string): string {
+    if (URL.canParse(pathOrUrl)) {
+      return new URL(pathOrUrl).pathname;
+    } else {
+      return resolve(this.basePath, pathOrUrl);
+    }
+  }
```
https://github.com/AztecProtocol/aztec-packages/issues/13069

spalladino committed:
Lint and add issue link

spalladino commented:
@alexghr I've addressed all comments or added issues, and renamed fast sync to snapshot (to avoid confusion with geth fast sync which is different). Other items remaining are:
- Cron this with a k8s job
- Add default for the snapshot url
- Add snapshot sync to _prover_ node

spalladino committed:
Allow downloading via http

spalladino committed:
Fixes.

spalladino committed:
Fix path to world state db version file

spalladino committed:
Fix unit tests

spalladino committed:
More fixes
